File: ecos-records-spring/src/main/java/ru/citeck/ecos/records2/spring/utils/registrar/RecordsDAORegistrar.java
package ru.citeck.ecos.records2.spring.utils.registrar;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ru.citeck.ecos.records2.RecordsService;
import ru.citeck.ecos.records2.source.dao.RecordsDAO;

import javax.annotation.PostConstruct;
import java.util.List;

@Slf4j
@Component
public class RecordsDAORegistrar {

    private RecordsService recordsService;
    private List<RecordsDAO> sources;

    @Autowired
    public RecordsDAORegistrar(RecordsService recordsService) {
        this.recordsService = recordsService;
    }

    @PostConstruct
    public void register() {
        log.info("========================== RecordsDAORegistrar ==========================");
        if (sources != null) {
            sources.forEach(this::register);
        }
        log.info("========================= /RecordsDAORegistrar ==========================");
    }

    private void register(RecordsDAO dao) {
        log.info("Register: \"" + dao.getId() + "\" with class " + dao.getClass().getName());
        recordsService.register(dao);
    }

    @Autowired(required = false)
    public void setSources(List<RecordsDAO> sources) {
        this.sources = sources;
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/RecordsDaoRegistrar.kt
package ru.citeck.ecos.records3

import io.github.oshai.kotlinlogging.KotlinLogging
import ru.citeck.ecos.records3.record.dao.RecordsDao

open class RecordsDaoRegistrar constructor(
    services: RecordsServiceFactory
) {

    companion object {
        val log = KotlinLogging.logger {}
    }

    private val recordsService: RecordsService = services.recordsService

    private var sources: List<RecordsDao>? = null

    fun register() {
        log.info { "========================== RecordsDaoRegistrar ==========================" }
        if (sources != null) {
            sources!!.forEach { this.register(it) }
        }
        log.info { "========================= /RecordsDaoRegistrar ==========================" }
    }

    private fun register(dao: RecordsDao) {
        log.info { "Register: \"" + dao.getId() + "\" with class " + dao.javaClass.name }
        recordsService.register(dao)
    }

    open fun setSources(sources: List<RecordsDao>?) {
        this.sources = sources
    }
}
